File: prac/Graph/Graph.cpp
#include "Graph.h"
#include <iostream>

Graph::Graph(bool d){
    directed = d;
    nvertices = 0; 
    nedges = 0;
    for(int i = 0; i <= MAXV; i++) degree[i] = 0;
}


void Graph::insert_edge(int x, int y, bool d){

    if(nedges > MAXOUTDEGREE) 
        cout << "Warning";
    
    edges[x][degree[x]] = y;
    degree[x] += 1;
    if(!d) insert_edge( y, x, true);
    else nedges++;
}

int* Graph::adj(int vertex){
    return edges[vertex];
}

void Graph::print_graph(void){
    for(int i = 0; i <= nvertices; i++){
        cout << i << " : ";
        for(int j = 0; j < degree[i]; j++){
            cout << edges[i][j] << " ";
        }   
        cout << "\n";
    }
}


void Graph::read_graph(void){
    int m;
    cin >> nvertices; //num of vertices
    cin >> m; //num of edges
    int x, y, d;
    //read in the edges
    for(int i = 0; i < m; i++){
        cin >> x;
        cin >> y;
        d = directed;
        insert_edge(x, y, d);
    }

}

File: prac/Graph/graph/lab8/graph.py
import sys
import os



class Graph:
    def __init__(self, v, directed=True):
        self.directed = directed
        self.nvertices = v
        self.nedges = v
        self.edges = [[0 for i in range(v)] for i in range(v)]
        self.degree = [0 for i in range(v)]

        


    def read_graph(self, filename):
        try:
            f = open(filename, 'r')
            data = f.readlines()
            #parse the data
            info = data[0]
            # the number of vertices in the graph
            self.nvertices = int(data[0].split()[0])
            n = int(data[0].split()[1]) # num of edges
            for i in range(1, n+1):
                x = int(data[i].split()[0])
                y = int(data[i].split()[1])
                self.insert_edge(x, y, self.directed)
    
        except IOError:
            print("Error reading in file")
            sys.exit()

    def insert_edge(self, x, y, directed):
        if x < self.nvertices:
            self.edges[x][y] = 1
            self.degree[x] += 1

        if directed != True:
            self.insert_edge(y, x, True)

    def print_graph(self):
        for edge in self.edges:
            for e in edge:
                print(e, " ", end="")
            print()

def main():
    g = Graph(10, False) 
    g.read_graph("graph.txt")
    g.print_graph()
if __name__ == "__main__":
    main()
    